marlowe/test/Spec/Marlowe/Semantics/Functions.hs

```diff
@@ -0,0 +1,961 @@
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.ValueGT`.
+checkValueGT :: Property
+checkValueGT =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval eval' _ _ (x, y) ->
+      eval' (ValueGT x y) == (eval x > eval y)
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.ValueLT`.
+checkValueLT :: Property
+checkValueLT =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval eval' _ _ (x, y) ->
+      eval' (ValueLT x y) == (eval x < eval y)
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.ValueLE`.
+checkValueLE :: Property
+checkValueLE =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval eval' _ _ (x, y) ->
+      eval' (ValueLE x y) == (eval x <= eval y)
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.ValueEQ`.
+checkValueEQ :: Property
+checkValueEQ =
+  let
+    gen _ _ = (,) <$> arbitrary <*> arbitrary
+  in
+    checkValue gen $ \eval eval' _ _ (x, y) ->
+      eval' (ValueEQ x y) == (eval x == eval y)
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.TrueObs`.
+checkTrueObs :: Assertion
+checkTrueObs =
+  assertBool "TrueObs is true."
+    $ evalObservation undefined undefined TrueObs
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.FalseObs`.
+checkFalseObs :: Assertion
+checkFalseObs =
+  assertBool "FalseObs is false."
+    . not $ evalObservation undefined undefined FalseObs
+
+
+-- | Test detection of actions not matching cases.
+checkApplyActionMismatch :: Property
+checkApplyActionMismatch = property $ do
+  let gen = do
+        let
+          inputs = [IDeposit undefined undefined undefined undefined, IChoice undefined undefined, INotify]
+          actions = [Deposit undefined undefined undefined undefined, Choice undefined undefined, Notify undefined]
+        x <- chooseInt (0, length inputs - 1)
+        y <- suchThat (chooseInt (0, length actions - 1)) (/= x)
+        pure (inputs !! x, actions !! y)
+  forAll' gen $ \(x, y) ->
+    case applyAction undefined undefined x y of
+      NotAppliedAction -> True
+      _                -> False
+
+
+-- | Test `Language.Marlowe.Core.V1.Semantics.Types.IDeposit` for a `Language.Marlowe.Core.V1.Semantics.Types.Deposit`.
+checkIDeposit :: Maybe Bool  -- ^ Whether the accounts match.
+              -> Maybe Bool  -- ^ Whether the parties match.
+              -> Maybe Bool  -- ^ Whether the tokens match.
+              -> Maybe Bool  -- ^ Whether the amounts match.
+              -> Property    -- ^ The test.
+checkIDeposit accountMatches partyMatches tokenMatches amountMatches = property $ do
+  let gen = do
+        context <- arbitrary
+        environment <- semiArbitrary context
+        state <- semiArbitrary context
+        (account, token) <- semiArbitrary context
+        accountMatches' <- maybe arbitrary pure accountMatches
+        account' <- if accountMatches' then pure account else suchThat arbitrary (/= account)
```

Comment:
Maybe this can be extracted as a function

Reply:
Thanks! Fixed as of 8e9144b8c75eecf8a9d7dd3e437bea6cbc686f65.